### PR TITLE
[groups persons] API for returning groups based on trend results

### PR DESCRIPTION
**Pull request by EDsCODE**
## Changes

*Please describe.*  
- breaking ground on adding ability to view groups when clicking on datapoint
- adds functionality for building a group query based on unique groups trend math
- moves test and adds persons querying to preexisting aggregate test
*If this affects the frontend, include screenshots.*  

## How did you test this code?

*Please describe.*
- moved/added test

### CODE REVIEW

**Comment by EDsCODE** [comment]
@macobo feel free to give it an initial pass to make sure it doesnt collide/violate any patterns you had in mind!

File: ee/clickhouse/queries/trends/person.py [hunk 1]
```diff
@@ -76,8 +82,39 @@ def get_query(self) -> Tuple[str, Dict]:
...
+            {**params, "offset": self.filter.offset, "limit": 200},
+        )
+
+    def get_query(self) -> Tuple[str, Dict]:
+        if self.is_group_query:
+            return self.get_group_query()
+        else:
+            return self.get_person_query()
+
     def get_people(self) -> ReturnDict:
```
Comment (macobo) [inline ee/clickhouse/queries/trends/person.py:109]: Can we add a TypedDict return type here?

**Review by macobo — CHANGES_REQUESTED** [review]
Left code-wise comments.

I am not a fan of the solution on a high level either though: It feels we've doubled the amount of queries we need to write and manage.

IMO the problems to solve are:
1. What to name groups + person query together? From this follows class/instance naming, endpoints, etc.
2. Can we unify the pattern for querying? E.g. something like this in a base class everything can use
```python
def get_query():
    event_query, event_params = ...
    if self.aggregating_by_groups: 
        query, params = self.groups_query()
        return query.format(...), { *event_params, *params }
    else:
        # same thing
```
    And have the `groups_query`/`person_query` be as standard as possible across all insights?
3. Can we unify the pattern for serializing? E.g. the return format is similar except for a few values and one literal which decides what we'll render in the FE. 
4. We should put a lot of effort into proper typing of return values.

I'm also confused at the lack of "links" along with datapoints in the code, but I might be missing context.

File: ee/clickhouse/views/test/test_clickhouse_trends.py [hunk 2]
```diff
@@ -332,3 +320,67 @@ def test_insight_trends_compare(self):
         assert sorted([p["id"] for p in prev_people]) == sorted(
             [str(created_people["p1"].uuid), str(created_people["p2"].uuid)]
         )
+
+
+class ClickhouseTestTrendsGroups(ClickhouseTestMixin, LicensedTestMixin, APIBaseTest):
```
[the remark this answers is not in the document]
Comment (EDsCODE) [reply ee/clickhouse/views/test/test_clickhouse_trends.py:325]: Moved test here because i changed it to an API test. Also, added a different class just so it's clearer and we can move it to a different file just for organizations sake

File: ee/clickhouse/views/test/test_clickhouse_trends.py [hunk 3]
```diff
@@ -332,3 +320,67 @@ def test_insight_trends_compare(self):
...
+
+        create_group(team_id=self.team.pk, group_type_index=0, group_key="org:5", properties={"industry": "finance"})
+        create_group(team_id=self.team.pk, group_type_index=0, group_key="org:6", properties={"industry": "technology"})
+        create_group(team_id=self.team.pk, group_type_index=0, group_key="org:7", properties={"industry": "finance"})
+        create_group(
+            team_id=self.team.pk, group_type_index=1, group_key="company:10", properties={"industry": "finance"}
+        )
+
+    @snapshot_clickhouse_queries
+    def test_aggregating_by_group(self):
```
[the remark this answers is not in the document]
Comment (EDsCODE) [reply ee/clickhouse/views/test/test_clickhouse_trends.py:341]: Do you mean testing the persons alongside the data or the multiple group type index filtering? Part of the improvements we were making last week was to connect the person querying with the data querying since the numbers should always match and that's relevant in testing them together

File: ee/clickhouse/sql/person.py [hunk 4]
```diff
@@ -317,3 +317,19 @@
...
 """
+
+GET_GROUPS_FROM_EVENT_QUERY = """
+SELECT 
+        {group_type_index},
+        $group_{group_type_index},
+        group_created_at_{group_type_index},
+        group_properties_{group_type_index}
+    FROM ({events_query})
+    GROUP BY
```
[the remark this answers is not in the document]
Comment (EDsCODE) [reply ee/clickhouse/sql/person.py:328]: I noticed this too and it exists in the trends query which was really confusing. I opted to just stick to almost exactly how the trend query is built for consistency. But now I can refactor them together

File: ee/clickhouse/sql/person.py [hunk 5]
```diff
@@ -317,3 +317,19 @@
...
+        group_created_at_{group_type_index},
+        group_properties_{group_type_index}
+    FROM ({events_query})
+    GROUP BY 
+        {group_type_index},
+        $group_{group_type_index},
+        group_created_at_{group_type_index},
+        group_properties_{group_type_index}
+ LIMIT 200
+OFFSET 0
```
[the remark this answers is not in the document]
Comment (EDsCODE) [reply ee/clickhouse/sql/person.py:334]: unfinished

File: ee/clickhouse/queries/groups_join_query.py [hunk 6]
```diff
@@ -1,9 +1,19 @@
...
 from ee.clickhouse.queries.column_optimizer import ColumnOptimizer
 from posthog.models import Filter
 from posthog.models.filters.path_filter import PathFilter
 from posthog.models.filters.retention_filter import RetentionFilter
+from posthog.models.property import GroupTypeIndex
+
+CREATED_AT = ColumnName("created_at")
+GROUP_PROPERTIES = ColumnName("group_properties")
+ACCEPTED_EXTRA_FIELDS = [CREATED_AT]
+GROUP_ALIASES = {
```
[the remark this answers is not in the document]
Comment (EDsCODE) [reply ee/clickhouse/queries/groups_join_query.py:13]: I added this type of query building so that [this select](https://github.com/PostHog/posthog/pull/7144/files#diff-68ee5fd398e73fa594adb5050743b2d33a3e9e407ca1444370432c49a18e6677R52) in the above query will be flexible and able to account for different fields of different group types. 

Also, if there's a query that's filtering on multiple group types, there will be multiple group joins so `Dict[GroupTypeIndex, ColumnName]` type allows you to specify which group join to add the fields in and for the query above to select the right one. 

Open to simplifying as you suggest but I think it would make the select in the outer query have more assumptions with regards to which group type index to select on

**Comment by EDsCODE** [comment]
Doesn't handle breakdown filtering yet but otherwise good for a review

File: ee/clickhouse/client.py [hunk 7]
```diff
@@ -145,8 +145,12 @@ def sync_execute(query, args=None, settings=None, with_column_types=False):
 
             try:
                 result = client.execute(
-                    prepared_sql, params=prepared_args, settings=settings, with_column_types=with_column_types
+                    prepared_sql,
+                    params=prepared_args,
+                    settings=settings,
+                    with_column_types=True if as_dict else with_column_types,
```
Comment (macobo) [inline ee/clickhouse/client.py:151]: `as_dict or with_column_types`

File: ee/clickhouse/client.py [hunk 8]
```diff
@@ -166,6 +170,15 @@ def sync_execute(query, args=None, settings=None, with_column_types=False):
                     save_query(prepared_sql, execution_time)
         return result
 
+    def _convert_to_dict(result) -> List[Dict[str, Any]]:
```
Comment (macobo) [inline ee/clickhouse/client.py:173]: This is pretty incomprehensible without better naming. Perhaps extract `header_row = result[1]` or smth?

File: ee/clickhouse/client.py [hunk 9]
```diff
@@ -135,7 +135,7 @@ def cache_sync_execute(query, args=None, redis_client=None, ttl=CACHE_TTL, setti
             redis_client.set(key, _serialize(result), ex=ttl)
             return result
 
-    def sync_execute(query, args=None, settings=None, with_column_types=False):
+    def sync_execute(query, args=None, settings=None, with_column_types=False, as_dict=False):
```
Comment (macobo) [inline ee/clickhouse/client.py:138]: Hmm. This muddies the return type of this base function a lot - it doesn't feel like we want even more options here.

Let's discuss this through during pairing.

File: ee/clickhouse/queries/actor_base_query.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        return res
+
+    def _serialize(self, data: Dict[str, Any]) -> Optional[ActorResponse]:
+        id = data.get("id", None)
+        if not id:
+            return None
+        created_at = data.get("created_at", None)
+        properties = json.loads(data.get("properties", "{}"))
+        is_identified: Optional[bool] = data.get("is_identified", None)
+        alias: str = properties.get("email", None) or properties.get("name", None)
```
Comment (macobo) [inline ee/clickhouse/queries/actor_base_query.py:61]: This only makes sense for persons, not groups. Groups have `group_key`

File: ee/clickhouse/queries/actor_base_query.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+class ActorBaseQuery:
+    aggregating_by_groups = False
+
+    def __init__(self, team: Team, filter: Filter, entity: Optional[Entity] = None):
+        self.team = team
+        self.entity = entity
+        self.filter = filter
+
+        if self.entity and self.entity.math == "unique_group":
```
Comment (macobo) [inline ee/clickhouse/queries/actor_base_query.py:27]: Comment got buried: I don't think this should be handled this way and instead either inlined or extracted into a property.

File: ee/clickhouse/queries/actor_base_query.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
+import json
+from typing import Any, Dict, List, Optional, Tuple, TypedDict
+
+from rest_framework import serializers
+
+from ee.clickhouse.client import sync_execute
+from posthog.models import Entity, Filter, Team
+
+
+class ActorResponse(TypedDict):
```
Comment (macobo) [inline ee/clickhouse/queries/actor_base_query.py:10]: This really isn't a `Response`, but only a subset of one.

Also there's no real way for the frontend to know given this object if it's dealing with a group or person, which will cause pain.

Rather than a single dict I think this would be much clearer as an union of two:

```
class SerializedPerson(TypedDict):
    type: Literal['person']
    ...

class SerializedGroup(TypedDict):
    type: Literal['group']
   ...

SerializedActor = Union[SerializedPerson, SerializedGroup]
```

File: ee/clickhouse/queries/groups_join_query.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (macobo) [reply ee/clickhouse/queries/groups_join_query.py:13]: > I think it would make the select in the outer query have more assumptions with regards to which group type index to select on

Let's take a look at the usecase:

```python
            join_queries.append(
                f"""
                INNER JOIN (
                    SELECT
                        group_key,
                        argMax(group_properties, _timestamp) AS group_properties_{group_type_index}
                        {(
                            " ".join(
                                f", argMax({field}, _timestamp) AS {GROUP_ALIASES[field]}{group_type_index}"
                                for field in columns_to_add
                            )
                        )}
                    FROM groups
                    WHERE team_id = %(team_id)s AND group_type_index = %({var})s
                    GROUP BY group_key
...
```

The problem I see it's this constant makes it becomes almost impossible to understand what this will expand to. One would need to:
- Jump to see what `_extra_fields` are passed
- Jump to see what `ACCEPTED_EXTRA_FIELDS` are, do the set logic
- Jump to constant to see what the mapping is
- Jump to query and see we're implicitly appending {group_type_index} at the end of it

We'd be jumping between so many different files for confusing reasons, making this hard to grok.

Let's discuss a potential solution. Perhaps we should be doing 

```
SELECT columns
FROM groups
WHERE id IN ({events_query})
```

instead and not bring this complexity within event queries and into all classes in the first place?

File: ee/clickhouse/queries/trends/person.py [hunk 13]
```diff
@@ -36,16 +38,21 @@ def _handle_date_interval(filter: Filter) -> Filter:
...
-        self.team = team
-        self.entity = entity
-        self.filter = filter
+class TrendsPersonQuery(ActorBaseQuery):
+    entity: Entity
 
-        if self.filter.display != TRENDS_CUMULATIVE and not self.filter.display in TRENDS_DISPLAY_BY_VALUE:
-            self.filter = _handle_date_interval(self.filter)
+    def __init__(self, team: Team, entity: Optional[Entity], filter: Filter):
+        new_filter = filter
```
Comment (macobo) [inline ee/clickhouse/queries/trends/person.py:45]: Nit: This temporary var isn't really needed

File: ee/clickhouse/queries/actor_base_query.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    id: str
+    created_at: Optional[str]
+    properties: Dict[str, Any]
+    is_identified: Optional[bool]
+    name: str
+    distinct_ids: List[str]
+
+
+class ActorBaseQuery:
```
Comment (macobo) [inline ee/clickhouse/queries/actor_base_query.py:19]: :+1: for this!

File: ee/clickhouse/views/test/test_clickhouse_trends.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (macobo) [reply ee/clickhouse/views/test/test_clickhouse_trends.py:341]: I mean it's doing two different `TrendsRequest`s below and on top of that has person logic following that.

Trying to do too many things in a test often causes it to become very hard to understand. Then when it breaks it gets deleted or "hacked around" rather than being properly handled.

File: posthog/api/utils.py [hunk 15]
```diff
@@ -19,16 +20,37 @@ class PaginationMode(Enum):
 
 
 def get_target_entity(request: request.Request) -> Entity:
```
Comment (macobo) [inline posthog/api/utils.py:22]: Q: What caused this change? I didn't see anything else relating to this changing.

File: ee/clickhouse/queries/groups_join_query.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (EDsCODE) [reply ee/clickhouse/queries/groups_join_query.py]: Resolved on call: We decided to always just retrieve IDs of groups or persons and retrieve them using postgres data. This way it's consistent across all insight implementations and doesn't require as much query reshaping to retrieve the right fields

File: ee/clickhouse/models/group.py [hunk 16]
```diff
@@ -2,12 +2,27 @@
...
 from ee.kafka_client.client import ClickhouseProducer
 from ee.kafka_client.topics import KAFKA_GROUPS
+from posthog.models.group import Group
+from posthog.utils import is_clickhouse_enabled
+
+if is_clickhouse_enabled():
+
+    @receiver(post_save, sender=Group)
+    def person_created(sender, instance: Group, created, **kwargs):
+        create_group(
```
Comment (EDsCODE) [inline ee/clickhouse/models/group.py:20]: mirrors the person implementation so that when groups are created from the django orm (especially during tests), the clickhouse and postgres data will match

File: ee/clickhouse/models/group.py [hunk 17]
```diff
@@ -2,12 +2,27 @@
...
 from ee.clickhouse.sql.groups import INSERT_GROUP_SQL
 from ee.kafka_client.client import ClickhouseProducer
 from ee.kafka_client.topics import KAFKA_GROUPS
+from posthog.models.group import Group
+from posthog.utils import is_clickhouse_enabled
+
+if is_clickhouse_enabled():
+
+    @receiver(post_save, sender=Group)
+    def person_created(sender, instance: Group, created, **kwargs):
```
Comment (macobo) [inline ee/clickhouse/models/group.py:19]: Nit: Name of the function is wrong.

We want to avoid this pattern if we can - this method won't be called in app code ever and if it is it's guaranteed to be wrong (e.g. properties_last_updated and properties_last_operation are missing), which will cause further bugs down the line.

I assume this is there for tests? If so, worth updating the `create_group` helper instead and perhaps update the name/file location to indicate this is test-specific?

File: ee/clickhouse/queries/column_optimizer.py [hunk 18]
```diff
@@ -104,7 +105,10 @@ def properties_used_in_filter(self) -> Counter[PropertyIdentifier]:
             #
             # See ee/clickhouse/queries/trends/util.py#process_math
             if entity.math == "unique_group":
-                counter[(f"$group_{entity.math_group_type_index}", "event", None)] += 1
+                if self.is_actor_query:
```
Comment (macobo) [inline ee/clickhouse/queries/column_optimizer.py:108]: Not sure this if-else is correct.

It's saying if it's an actor query, property `$group_{N}` for group type math_group_type_index is used, which does not seem correct.

File: ee/clickhouse/queries/column_optimizer.py [hunk 19]
```diff
@@ -19,9 +19,10 @@ class ColumnOptimizer:
     This speeds up queries since clickhouse ends up selecting less data.
     """
 
-    def __init__(self, filter: Union[Filter, PathFilter, RetentionFilter], team_id: int):
+    def __init__(self, filter: Union[Filter, PathFilter, RetentionFilter], team_id: int, is_actor_query: bool = False):
```
Comment (macobo) [inline ee/clickhouse/queries/column_optimizer.py:22]: Q: Why is this needed?

In general we want to avoid coupling column_optimizer with other modules even tighter. Since everything already depends on it, we end up with a two-way dependency which is just hell to debug.

File: ee/clickhouse/queries/event_query.py [hunk 20]
```diff
@@ -103,11 +104,6 @@ def _determine_should_join_persons(self) -> None:
             self._should_join_persons = True
             return
 
-        if self._filter.breakdown_type == "person":
```
Comment (macobo) [inline ee/clickhouse/queries/event_query.py:106]: Q: Why remove this? Was this handled by column_optimizer already?

File: ee/clickhouse/queries/column_optimizer.py (hunk 19 above)
Comment (EDsCODE) [reply ee/clickhouse/queries/column_optimizer.py:22]: Didn't like the coupling either but seemed like the most streamlined way to add in the flag that would tell the querybuilder to add the group join in

File: ee/clickhouse/queries/actor_base_query.py [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+            if isinstance(person, Person)
+        ]
+
+    def _serialize_groups(self, data: QuerySet[Actor]) -> List[SerializedActor]:
+        return [
+            SerializedGroup(
+                type="group", group_key=group.group_key, created_at=group.created_at, properties=group.group_properties
+            )
+            for group in data
+            if isinstance(group, Group)
```
Comment (macobo) [inline ee/clickhouse/queries/actor_base_query.py:113]: Can it ever not be Group here? I think not right?

If the goal is type-casting then I'd change the signature to

```
    def _serialize_groups(self, data: QuerySet[Group]) -> List[SerializedActor]:
```

File: ee/clickhouse/queries/actor_base_query.py [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+        return actors, serialized_actors
+
+    def _get_groups(self, results) -> Tuple[QuerySet[Actor], List[SerializedActor]]:
+        """ Get groups from raw SQL results in data model and dict formats """
+        groups: QuerySet[Actor] = Group.objects.filter(team_id=self.team.pk, group_key__in=[val[0] for val in results])
+        return groups, self._serialize_groups(groups)
+
+    def _get_people(self, results) -> Tuple[QuerySet[Actor], List[SerializedActor]]:
+        """ Get people from raw SQL results in data model and dict formats """
+        persons: QuerySet[Actor] = Person.objects.filter(team_id=self.team.pk, uuid__in=[val[0] for val in results])
```
Comment (macobo) [inline ee/clickhouse/queries/actor_base_query.py:87]: See below comment on the _serialize_people signature. If you annotate this as `persons: QuerySet[Person]` everything should work well.

File: ee/clickhouse/queries/actor_base_query.py [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+
+    def __init__(self, team: Team, filter: Filter, entity: Optional[Entity] = None):
+        self.team = team
+        self.entity = entity
+        self.filter = filter
+
+        if self.entity and self.entity.math == "unique_group":
+            self.aggregating_by_groups = True
+
+    def groups_query(self) -> Tuple[str, Dict]:
```
Comment (macobo) [inline ee/clickhouse/queries/actor_base_query.py:52]: Q: Why do subclasses need to implement both? Existing queries are already using `aggregation_target` as the column to return from event_query, which would be the group key in one case and person_id in the other.

Comment (macobo) [reply ee/clickhouse/queries/actor_base_query.py:52]: That said, looking at the only implementation these are really similar. I'll need to read through the code with a fresh head tomorrow, I _think_ there might be ways to simplify this further.

File: ee/clickhouse/queries/event_query.py (hunk 20 above)
Comment (EDsCODE) [reply ee/clickhouse/queries/event_query.py]: no idea i think it got deleted when i merged in your branch because of previously edited deletions

File: ee/clickhouse/views/test/test_clickhouse_trends.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (EDsCODE) [reply ee/clickhouse/views/test/test_clickhouse_trends.py:348]: removed second part

File: ee/clickhouse/queries/column_optimizer.py (hunk 19 above)
Comment (EDsCODE) [reply ee/clickhouse/queries/column_optimizer.py]: nvm. was looking at this the wrong way. after we decided to use postgres data, very little changes were needed

File: ee/clickhouse/queries/actor_base_query.py (hunk 23 above)
Comment (EDsCODE) [reply ee/clickhouse/queries/actor_base_query.py]: Yup you're right. I combined the functions

File: ee/clickhouse/queries/trends/person.py [hunk 24]
```diff
@@ -74,19 +80,20 @@ def get_query(self) -> Tuple[str, Dict]:
             filter=self.filter,
             team_id=self.team.pk,
             entity=self.entity,
-            should_join_distinct_ids=True,
-            should_join_persons=True,
-            extra_fields=["distinct_id", "team_id"],
-            extra_person_fields=["created_at", "person_props", "is_identified"],
+            should_join_distinct_ids=False if self.is_aggregating_by_groups else True,
+            should_join_persons=False if self.is_aggregating_by_groups else True,
+            extra_fields=[] if self.is_aggregating_by_groups else ["distinct_id", "team_id"],
```
Comment (EDsCODE) [inline ee/clickhouse/queries/trends/person.py:85]: Realized this could just be a single query after we agreed we can just pull data from postgres with the ids. There's very few extra params needed when it's a group query besides these 3. No need to push an extra parameters down to columnoptimizer or the eventquery builder

File: posthog/api/utils.py (hunk 15 above)
Comment (EDsCODE) [reply posthog/api/utils.py:22]: Retrieving groups from trends test was failing because this helper function was building the entity using solely ENTITY_TYPE and ENTITY_MATH but didn't have the group type index

**Review by EDsCODE — COMMENTED** [review]
Ended up needing very few changes. Most of the diffs are from tests and snapshot changes

File: ee/clickhouse/models/group.py (hunk 17 above)
Comment (EDsCODE) [reply ee/clickhouse/models/group.py]: Hmm I added the test flag to it. Not quite sure where to move this logic though where it will still be obvious

File: ee/clickhouse/queries/actor_base_query.py [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+        self.entity = entity
+        self.filter = filter
+
+    def actor_query(self) -> Tuple[str, Dict]:
+        """ Implemented by subclasses. Must return list of uuids. Can be group uuids (group_key) or person uuids """
+        raise NotImplementedError()
+
+    @cached_property
+    def is_aggregating_by_groups(self) -> bool:
+        if self.entity and self.entity.math == "unique_group":
```
Comment (macobo) [inline ee/clickhouse/queries/actor_base_query.py:57]: nit: `return self.entity is not None and self.entity.math == "unique_group"`

File: ee/clickhouse/queries/groups_join_query.py [hunk 26]
```diff
@@ -4,6 +4,7 @@
 from posthog.models import Filter
 from posthog.models.filters.path_filter import PathFilter
 from posthog.models.filters.retention_filter import RetentionFilter
+from posthog.models.property import GroupTypeIndex
```
Comment (macobo) [inline ee/clickhouse/queries/groups_join_query.py:7]: Nit: Dead import after changes.

File: ee/clickhouse/queries/event_query.py [hunk 27]
```diff
@@ -1,5 +1,5 @@
 from abc import ABCMeta, abstractmethod
-from typing import Any, Dict, List, Tuple, Union
+from typing import Any, Dict, List, Optional, Set, Tuple, Union
```
Comment (macobo) [inline ee/clickhouse/queries/event_query.py:2]: Nit: There's some new dead imports here

File: ee/clickhouse/queries/trends/person.py [hunk 28]
```diff
@@ -74,19 +80,20 @@ def get_query(self) -> Tuple[str, Dict]:
             filter=self.filter,
             team_id=self.team.pk,
             entity=self.entity,
-            should_join_distinct_ids=True,
-            should_join_persons=True,
-            extra_fields=["distinct_id", "team_id"],
-            extra_person_fields=["created_at", "person_props", "is_identified"],
+            should_join_distinct_ids=False if self.is_aggregating_by_groups else True,
+            should_join_persons=False if self.is_aggregating_by_groups else True,
```
Comment (macobo) [inline ee/clickhouse/queries/trends/person.py:84]: Nit: `should_join_persons=not self.is_aggregating_by_groups` is imo cleaner :)

File: ee/clickhouse/queries/trends/person.py [hunk 29]
```diff
@@ -36,16 +39,19 @@ def _handle_date_interval(filter: Filter) -> Filter:
     return filter.with_data(data)
 
 
-class TrendsPersonQuery:
-    def __init__(self, team: Team, entity: Entity, filter: Filter):
-        self.team = team
-        self.entity = entity
-        self.filter = filter
+class TrendsPersonQuery(ActorBaseQuery):
```
Comment (macobo) [inline ee/clickhouse/queries/trends/person.py:42]: Note: We should rename this + endpoints in future PRs as well.

File: ee/clickhouse/sql/person.py [hunk 30]
```diff
@@ -299,21 +299,11 @@
...
-SELECT
-    person_id,
-    created_at,
-    team_id,
-    person_props,
-    is_identified,
-    arrayReduce('groupUniqArray', groupArray(distinct_id)) AS distinct_ids
+GET_ACTORS_FROM_EVENT_QUERY = """
+SELECT 
+    {id_field} AS id
```
Comment (macobo) [inline ee/clickhouse/sql/person.py:304]: Nit: let's name this `AS actor_id` - this way we can be consistent in other queries. id is IMO overloaded.

**Review by macobo — APPROVED** [review]
:heart_eyes: This took some iteration but really happy where this ended up - I think this is quite a sustainable pattern for the future.

File: ee/clickhouse/models/group.py [hunk 31]
```diff
@@ -2,12 +2,28 @@
...
 from ee.clickhouse.sql.groups import INSERT_GROUP_SQL
 from ee.kafka_client.client import ClickhouseProducer
 from ee.kafka_client.topics import KAFKA_GROUPS
+from posthog.models.group import Group
+from posthog.utils import is_clickhouse_enabled
+
+if is_clickhouse_enabled() and settings.TEST:
+
+    @receiver(post_save, sender=Group)
+    def group_created(sender, instance: Group, created, **kwargs):
```
Comment (macobo) [inline ee/clickhouse/models/group.py:20]: I'd move this inside the `create_group` function. It's only ever used in tests.

File: ee/clickhouse/queries/actor_base_query.py (hunk 25 above)
Comment (EDsCODE) [reply ee/clickhouse/queries/actor_base_query.py:57]: missed this but added it in the funnel PR
